projects/plugins/crm/src/automation/class-automation-engine.php

```diff
@@ -0,0 +1,191 @@
+<?php
+
+namespace Automattic\Jetpack\CRM\Automation;
+
+/**
+ * Automation Engine
+ *
+ * @package Automattic\Jetpack\CRM\Automation
+ */
+class Automation_Engine {
+
+	/** @var Automation_Engine Instance singleton */
+	private static $instance = null;
+
+	/** @var array triggers map name => classname */
+	private $triggers_map = array();
+	
+	/** @var array steps map name => classname */
+	private $steps_map = array();
+
+	/** @var Automation_Logger Automation logger */
+	private $automation_logger;
+
+	/** @var array */
+	private $workflows = array();
+
+	/**
+	 *  Instance singleton object
+	 *
+	 * @return Automation_Engine
+	 */
+	public static function instance(): Automation_Engine {
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Set the automation logger
+	 *
+	 * @param Automation_Logger $logger The automation logger.
+	 */
+	public function set_automation_logger( Automation_Logger $logger ) {
+		$this->automation_logger = $logger;
+	}
+
+	/**
+	 * Register a trigger
+	 *
+	 * @param string Trigger $trigger_name
+	 * @param string         $class_name
+	 * @throws Automation_Exception
+	 */
+	public function register_trigger( string $trigger_name, string $class_name ) {
```

Comment:
Should we try to simplify this slightly and find the `$trigger_name` value on the trigger itself?

Reply:
We are mapping/registering classes with the name that we will identify the trigger.

`"contact_updated" => "Automattic\Jetpack\CRM\Automation\Triggers\Contact_Updated"`

That "simplification" will imply instancing the trigger only to get the name. Or, setting the `name` as a static field.

The part of how we are going to discover triggers, actions, and conditions, is not totally defined yet. But my first idea is that probably all those fields (name, title, category,...) will become static fields or constants. We don't need to instance them until we really need them.